edb/server/protocol/auth_ext/otc.py

```diff
@@ -0,0 +1,238 @@
+
+    Args:
+        db: Database connection
+        factor_id: The ID of the factor to associate with this code
+        ttl: Time to live for the code
+
+    Returns:
+        Tuple of (code, otc_id) where code is the plain text code and otc_id is
+        the database ID
+    """
+    code = generate_code()
+    code_hash = hash_code(code)
+    expires_at = datetime.datetime.now(datetime.timezone.utc) + ttl
+
+    r = await execute.parse_execute_json(
+        db=db,
+        query="""\
+with
+    ONE_TIME_CODE := (insert ext::auth::OneTimeCode {
+        factor := <ext::auth::Factor><uuid>$factor_id,
+        code_hash := <bytes>$code_hash,
+        expires_at := <datetime>$expires_at,
+    })
+select ONE_TIME_CODE { id };""",
+        variables={
+            "factor_id": factor_id,
+            "code_hash": code_hash,
+            "expires_at": expires_at.isoformat(),
+        },
+        cached_globally=True,
+        query_tag='gel/auth',
+    )
+
+    result_json = json.loads(r.decode())
+    if len(result_json) != 1:
+        raise errors.InvalidData("Failed to create OneTimeCode")
+
+    return code, uuid.UUID(result_json[0]["id"])
+
+
+async def verify(db: Any, factor_id: str, code: str) -> str:
+    """
+    Verify a one-time code for a given factor.
+
+    This function performs all verification operations in a single atomic query:
+    - Cleanup expired codes
+    - Check rate limits
+    - Find and validate the code
+    - Delete the code if valid
+    - Record the authentication attempt
+
+    Args:
+        db: Database connection
+        factor_id: The ID of the factor
+        code: The code to verify
+
+    Returns:
+        The OneTimeCode ID if verification succeeds
+
+    Raises:
+        NoIdentityFound: If code is invalid, expired, or max attempts exceeded
+    """
+    code_hash = hash_code(code)
+
+    # Single composed query that handles the entire verification workflow
+    r = await execute.parse_execute_json(
+        db=db,
+        query="""\
+with
+    factor_id := <uuid>$factor_id,
+    FACTOR := (select ext::auth::Factor filter .id = factor_id),
+    code_hash := <bytes>$code_hash,
+    now := datetime_current(),
+    window_start := now - <duration>'10 minutes',
+
+    # Cleanup expired codes (side effect)
+    cleanup_count := count(
+        delete ext::auth::OneTimeCode filter .expires_at < now
+    ),
+
+    # Check rate limits
+    failed_attempts := (
+        select count(
+            select ext::auth::AuthenticationAttempt
+            filter .factor = FACTOR
+               and .attempt_type =
+                   ext::auth::AuthenticationAttemptType.OneTimeCode
+               and .successful = false
+               and .created_at > window_start
+        )
+    ),
+
+    # Find the OTC
+    otc := (
+        select ext::auth::OneTimeCode
+        filter .factor = FACTOR and .code_hash = code_hash
+        limit 1
+    ),
+
+    is_rate_limited := failed_attempts >= 5,
```

Comment:
Make `5` either a query parameter or with bind it separately as a "constant"?

Reply:
Yeah, I actually want to make this configurable at some point in the future, but for now, I'll bind it as a constant, good idea.